src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1658,22 +1718,134 @@ protected override void VisitArguments(ImmutableArray<BoundExpression> arguments
+            ImmutableArray<int> argsToParamsOpt,
+            bool expanded)
+        {
+            Debug.Assert(!IsConditionalState);
+
+            ImmutableArray<(bool notNullWhenFalse, bool ensuresNotNull)> annotations = GetAnnotations(arguments.Length,
+               expanded, parameters, argsToParamsOpt);
+
+            if (annotations.IsDefault)
+            {
+                return;
+            }
+
+            bool saveDisableDiagnostics = _disableDiagnostics;
+            _disableDiagnostics = true;
+
+            Debug.Assert(annotations.Length == arguments.Length);
+
+            // Since we may be splitting states to represent the method returning true/false,
+            // we'll prepare all the slots ahead of time.
+            var slots = ArrayBuilder<int?>.GetInstance(arguments.Length);
+            for (int i = 0; i < arguments.Length; i++)
+            {
                 var argument = arguments[i];
-                if (refKind != RefKind.Out)
+                int slot = MakeSlot(argument);
+                if (slot <= 0)
+                {
+                    slots.Add(null);
+                    continue;
+                }
+                if (slot >= this.State.Capacity) Normalize(ref this.State);
+                slots.Add(slot);
+            }
+
+            for (int i = 0; i < arguments.Length; i++)
+            {
+                if (this.IsConditionalState)
                 {
-                    // PROTOTYPE(NullReferenceTypes): `ref` arguments should be treated as l-values
-                    // for assignment. See `ref x3` in StaticNullChecking.PassingParameters_01.
-                    VisitRvalue(argument);
+                    // We could be in a conditional state because of NotNullWhenFalse annotation
+                    // Then WhenTrue/False states correspond to the invocation returning true/false
+
+                    // We'll successively assume that we're in the unconditional state where the method returns true,
+                    // then we'll repeat assuming the method returns false.
+
+                    LocalState whenTrue = this.StateWhenTrue.Clone();
+                    LocalState whenFalse = this.StateWhenFalse.Clone();
+
+                    this.SetState(whenTrue);
+                    VisitArgumentEvaluate(arguments, refKindsOpt, i);
+                    Debug.Assert(!IsConditionalState);
+
+                    this.SetState(whenFalse);
+                    VisitArgumentEvaluate(arguments, refKindsOpt, i);
+                    Debug.Assert(!IsConditionalState);
+
+                    this.SetConditionalState(whenTrue, whenFalse);
                 }
                 else
                 {
-                    VisitLvalue(argument);
+                    VisitArgumentEvaluate(arguments, refKindsOpt, i);
+                }
+
+                var argument = arguments[i];
+                if (argument.Type?.IsReferenceType != true || slots[i] == null)
```

Comment:
We may need to handle unconstrained type parameters as well: `AssertNotNull<T>([EnsuresNotNull] T t);`. Perhaps add a `// PROTOTYPE` comment.

Reply:
I'll be adding a test (`EnsuresNotNull_GenericMethod`). It just worked though. Lucky

---
In reply to: [187071624](https://github.com/dotnet/roslyn/pull/26656#discussion_r187071624) [](ancestors = 187071624)